Create snapshot-based instance restore feature

hyunihs opened the pull request:
## Snapshot-based instance restore feature  
- Closes #14 

#### Implement `delete` command with snapshot creation (`dumie delete <profile>`)  
- Find EC2 instance by profile name  
- Extract root EBS volume  
- Create a snapshot tagged with the profile name and instance ID
- Terminate instance 
- Delete outdated snapshots and AMIs

#### Add snapshot-based restore on `dumie deploy manual <profile>`  
- Look up snapshots tagged with the given profile  
- Register an AMI from the snapshot  
- Launch a new EC2 instance from the AMI (snapshot)
- If no snapshot is found, launch a new instance using the latest Amazon Linux AMI

chanhyeokseo commented:
Hey @hyunihs , thanks for opening the PR and nice work implementing the snapshot & delete feature.

Regarding the snapshot command, I designed this feature to have only two commands: deploy and delete.

When you're creating the instance, it will automatically detect and restore the snapshot if exists. The identifier is the profile name here.

Your implementation seems it is using a separate snapshot command, which will add an extra command for users to remember. 

Any insights would be appreciated. Thanks!

hyunihs commented:
> Hey @hyunihs , thanks for opening the PR and nice work implementing the snapshot & delete feature.
> 
> Regarding the snapshot command, I designed this feature to have only two commands: deploy and delete.
> 
> When you're creating an instance, it will automatically detect and restore the snapshot if exists. The identifier here is the profile name.
> 
> Your implementation seems it is using a separate snapshot command, which will add an extra command for users to remember.
> 
> Any insights would be appreciated. Thanks!

Thank you for the feedback — I agree that restoring from a snapshot should be a part of the `deploy` flow rather than a separate command.

Based on your suggestion, I made the following changes:

- Removed the `snapshot` subcommand.
- Integrated snapshot-based restoration directly into the `deploy <profile>` flow. When a user runs `dumie deploy manual <profile>`, Dumie checks for a snapshot tagged with the profile name and restores from it if available. Otherwise, it falls back to launching a new instance.
- Since this logic will be shared across multiple subcommands (`manual`, `schedule`, `ttl`, etc.), I extracted the snapshot restoration and instance creation flow (along with lock handling) into a centralized module (`deployment.go`) for reuse and consistency.
- Updated the `delete <instance-id>` command to `delete <profile>` to align with the rest of the profile-based CLI interface.

I've also updated the PR description to reflect these changes.

chanhyeokseo commented:
Hi @hyunihs , thanks for applying those changes : )

I think we should destroy the previously existing snapshot before creating a new snapshot.

Currently, it will create a new snapshot whenever we delete an instance. For that reason, dumie cannot choose what to restore when creating the instance again.

I've tested your code locally and found that it actually restoring from the old snapshot.

I think we have two options:
1. Restore snapshots based on a timestamp (in case users want to restore older ones)
2. Delete any pre-existing snapshot when deleting an instance (keeping only the latest)

Let me know what you think, and feel free to reach out if you need help on my end!

chanhyeokseo commented on dumie-cli/awsutils/ec2/snapshot.go:
```diff
@@ -0,0 +1,104 @@
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []string{profile},
+			},
```
minor change to add another filter to find from the ManagedBy:Dumie tag (since you've added this tag when you're creating a snapshot)

```
			{
				Name:   aws.String("tag:ManagedBy"),
				Values: []string{"Dumie"},
			},
```

would be better to prevent any collisions

chanhyeokseo commented on dumie-cli/awsutils/ec2/deployment.go:
```diff
@@ -0,0 +1,82 @@
+		return "", fmt.Errorf("failed to get security group: %w", err)
+	}
+
+	keyName, err := common.GenerateKeyPair(client)
+	if err != nil {
+		return "", fmt.Errorf("failed to get key pair: %w", err)
+	}
```
```suggestion
	keyName, err := common.GetKeyPairName()
	if err != nil {
		return "", fmt.Errorf("error getting key pair name: %v", err)
	}
```
your code works, however, we have a separate function to get the key pair name.

using `GenerateKeyPair` directly might be risky because when there's no key pair associated in your local env it will create a new one and **will not save it to the config file**.

I recommend using `GetKeyPair` instead, but if you indeed want to use this function here, you need to consider saving the new key name info to the config file.

chanhyeokseo replied:
```suggestion
	keyName, err := common.GetKeyPair(client)
	if err != nil {
		return "", fmt.Errorf("failed to get key pair: %w", err)
	}
```

same here (see above comment)

chanhyeokseo commented on dumie-cli/awsutils/ec2/snapshot.go:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	// Get Pem Key
+	keyName, err := common.GenerateKeyPair(client)
+	if err != nil {
+		return "", fmt.Errorf("Error getting key pair name: %v\n", err)
+	}
```
```suggestion
	keyName, err := common.GetKeyPairName()
	if err != nil {
		return "", fmt.Errorf("error getting key pair name: %v", err)
	}
```
your code works, however, we have a separate function to get the key pair name.

using `GenerateKeyPair` directly might be risky because when there's no key pair associated in your local env it will create a new one and **will not save it to the config file**.

I recommend using `GetKeyPair` instead, but if you indeed want to use this function here, you need to consider saving the new key name info to the config file.

hyunihs replied:
I've switched to using `GetKeyPair`, and added a more descriptive error message to guide users to run `dumie configure` if the key pair isn't set up yet.
Thanks for the suggestion 👍

chanhyeokseo commented:
also i think we need to delete AMIs when deleting instances?

hyunihs commented:
Thanks for the feedback!

- When deleting an instance, Dumie now keeps only the latest snapshot and deletes outdated snapshots (along with AMIs) after termination. This ensures the next `deploy` always restores from the most recent snapshot.
- Replaced `GenerateKeyPair` with `GetKeyPair` when launching instances.
- Improved the error message shown when the key pair is missing, to guide users to run `dumie configure` first.
- Added a `ManagedBy: Dumie `tag filter when deleting snapshots to avoid affecting unrelated ones.

Let me know if there's anything else to improve!

chanhyeokseo commented:
Hi @hyunihs ,

The code seems good to me so thanks for the contribution 👍 

I'm merging it now
